src/skagent/algos/maliar.py

```diff
@@ -308,4 +308,183 @@ def maliar_training_loop(
         # TODO: test for difference.. how? This effects the FOR (/while) loop above.
 
     # Step 3. Assess the accuracy of constructed approximation ϕ (·, θ ) on a new sample.
-    return ann, states
+    return bpn, states
+
+
+def estimate_bellman_residual(
+    block,
+    discount_factor,
+    value_network,
+    df,
+    states_t,
+    shocks_t,
+    parameters={},
+    agent=None,
+):
+    """
+    Computes the Bellman equation residual for given states and shocks.
+
+    The Bellman equation is: V(s) = max_c { u(s,c) + β E[V(s')] }
+    This function computes: V(s) - [u(s,c) + β V(s')]
+
+    Parameters
+    ----------
+    block : model.DBlock
+        The model block containing dynamics, rewards, and shocks
+    discount_factor : float
+        The discount factor β
+    value_network : callable
+        A value function that takes state variables and returns value estimates
+    df : callable
+        Decision function that returns controls given states and shocks
+    states_t : dict
+        Current state variables
+    shocks_t : dict
+        Current shock realizations
+    parameters : dict, optional
+        Model parameters for calibration
+    agent : str, optional
+        Agent identifier for rewards
+
+    Returns
+    -------
+    torch.Tensor
+        Bellman equation residual
+    """
+    if callable(discount_factor):
+        raise Exception(
+            "Currently only numerical, not state-dependent, discount factors are supported."
+        )
+
+    # Get state variable names for transition
+    state_variables = list(states_t.keys())
+
+    # Get reward variables
+    reward_vars = [
+        sym for sym in block.reward if agent is None or block.reward[sym] == agent
+    ]
+    if len(reward_vars) == 0:
+        raise Exception("No reward variables found in block")
+    reward_sym = reward_vars[0]  # Assume single reward for now
+
+    # Get current value estimates
+    current_values = value_network(states_t, shocks_t, parameters)
+
+    # Get controls from decision function
+    controls_t = df(states_t, shocks_t, parameters)
+
+    # Create transition and reward functions
+    tf = create_transition_function(block, state_variables)
+    rf = create_reward_function(block, agent)
+
+    # Compute immediate reward
+    immediate_reward = rf(states_t, shocks_t, controls_t, parameters)[reward_sym]
+
+    # Compute next states
+    next_states = tf(states_t, shocks_t, controls_t, parameters)
+
+    # Compute continuation value using value network
+    continuation_values = value_network(next_states, shocks_t, parameters)
```

Comment:
It looks to me like the `shocks_t` input is doing the work of the shocks in period `t` (in the calculation of the 'immediate reward') as well `t+1` in the computation of the continuation value (when passed to the value network).

I'm surprised by this. Do we actually need two different samples of the shocks here, so we have independent draws for the `t` and `t+1` shocks at work?

Reply:
Since automated tests were passing despite this being a bug, there should be an automated test (maybe from one of the benchmarks) that tests for for the solver's performance on a problem with genuine time-dependence.